engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ActivateProcessInstanceBatchProcessor.java

```diff
@@ -0,0 +1,104 @@
+  }
+
+  @Override
+  public void processRecord(final TypedRecord<ProcessInstanceBatchRecord> record) {
+    final var recordValue = record.getValue();
+
+    final ProcessInstanceRecord childInstanceRecord = createChildInstanceRecord(recordValue);
+    var amountOfChildInstancesToActivate = recordValue.getIndex();
+    while (amountOfChildInstancesToActivate > 0) {
+      if (canWriteCommands(record, childInstanceRecord)) {
+        final long childInstanceKey = keyGenerator.nextKey();
+        commandWriter.appendFollowUpCommand(
+            childInstanceKey, ProcessInstanceIntent.ACTIVATE_ELEMENT, childInstanceRecord);
+        amountOfChildInstancesToActivate--;
+      } else {
+        writeFollowupBatchCommand(recordValue, amountOfChildInstancesToActivate);
+        break;
+      }
+    }
+  }
+
+  private ProcessInstanceRecord createChildInstanceRecord(
+      final ProcessInstanceBatchRecord recordValue) {
+    final var parentElementInstance =
+        elementInstanceState.getInstance(recordValue.getBatchElementInstanceKey());
+    final var processDefinition =
+        processState
+            .getProcessByKey(parentElementInstance.getValue().getProcessDefinitionKey())
+            .getProcess();
+
+    final var parentElement =
+        processDefinition.getElementById(parentElementInstance.getValue().getElementId());
+    final var childElement = ((ExecutableMultiInstanceBody) parentElement).getInnerActivity();
+
+    final var childInstanceRecord = new ProcessInstanceRecord();
+    childInstanceRecord.wrap(parentElementInstance.getValue());
+    childInstanceRecord
+        .setFlowScopeKey(parentElementInstance.getKey())
+        .setElementId(childElement.getId())
+        .setBpmnElementType(childElement.getElementType())
+        .setBpmnEventType(childElement.getEventType());
+    return childInstanceRecord;
+  }
+
+  private void writeFollowupBatchCommand(
+      final ProcessInstanceBatchRecord recordValue, final long index) {
+    final var nextBatchRecord =
+        new ProcessInstanceBatchRecord()
+            .setProcessInstanceKey(recordValue.getProcessInstanceKey())
+            .setBatchElementInstanceKey(recordValue.getBatchElementInstanceKey())
+            .setIndex(index);
+    final long key = keyGenerator.nextKey();
+    commandWriter.appendFollowUpCommand(key, ProcessInstanceBatchIntent.ACTIVATE, nextBatchRecord);
+  }
+
+  private boolean canWriteCommands(
+      final TypedRecord<ProcessInstanceBatchRecord> record,
+      final ProcessInstanceRecord childInstanceRecord) {
+    // We must have space in the batch to write both the ACTIVATE command as the potential
+    // follow-up batch command. An excessive 8Kb is added to account for metadata. This is way
+    // more than will be necessary.
+    final var expectedCommandLength =
+        record.getLength() + childInstanceRecord.getLength() + (1024 * 8);
+    return commandWriter.canWriteCommandOfLength(expectedCommandLength);
+  }
+}
```

Comment:
:thought_balloon: I wonder if it makes sense to extract these two methods in a `*Behavior` class since there are two somewhat similar methods in the `TerminateProcessInstanceBatchProcessor`, and I assume we might have further use of the `ProcessInstanceBatch` record in the future, so we might duplicate these methods even more.

Reply:
I was thinking about this as well. Thing is, in the end it's only 2 lines of code. One of which is used to determine the `expectedCommandLength`. This is different for both of the processors we have now. Moving this to a behavior class would mean that we require a parameter getting the `expectedCommandLength`. So in the end all the behavior method will do make a call to the command writer.

💭 I guess we could argue that the + 8KB is duplicated and could be extracted from the calculation. To be honest, at this time I don't really care. Let's see if we start using this command more often and how similar the other checks will become.